### PR TITLE
Implement AMQP request-response matcher

### PR DESCRIPTION
**Pull request by mertyildiran**
While I tried to run the front-end locally, I got missing types error so I ran `npm install --save-dev @types/ace` that's why there is a diff in front-end code.

Screenshot from 2022-07-08 23-06-31
Screenshot from 2022-07-08 23-06-39
Screenshot from 2022-07-08 23-06-41
Screenshot from 2022-07-08 23-06-53
Screenshot from 2022-07-08 23-06-57

### CODE REVIEW

**Comment by gadotroee** [comment]
@mertyildiran, can this change help with the acceptance tests or is it not related? 

I will take a look later today or tomorrow morning (I know it waited for some time)

File: tap/extensions/amqp/helpers.go [hunk 1]
```diff
@@ -593,6 +634,65 @@ func representConnectionStart(event map[string]interface{}) []interface{} {
...
+		for name, value := range event["clientProperties"].(map[string]interface{}) {
+			var outcome string
+			switch v := value.(type) {
+			case string:
+				outcome = v
+			case map[string]interface{}:
+				x, _ := json.Marshal(value)
+				outcome = string(x)
+			default:
+				panic("Unknown data type for the client property!")
```
Comment (gadotroee) [inline tap/extensions/amqp/helpers.go:674]: I'm not sure we want to panic (that can make the whole api server to restart) here, maybe just return empty representation and add some log?

Comment (mertyildiran) [reply tap/extensions/amqp/helpers.go]: Fixed with https://github.com/up9inc/mizu/commit/586046150d98fa748966f502443862ffa8083324

**Comment by mertyildiran** [comment]
> @mertyildiran, can this change help with the acceptance tests or is it not related?
> 
> I will take a look later today or tomorrow morning (I know it waited for some time)
> 
> I see that you ran the acceptance tests but it will not run the AMQP acceptance because of this (https://github.com/up9inc/mizu/blob/develop/acceptanceTests/extensions_test.go#L113). We disabled it in the past when they were not stable - so if it can help with the tests we might need to remove this

I enabled the acceptance tests for AMQP with https://github.com/up9inc/mizu/commit/c60e9cf747c0dccc45345790c2c9f49311a0278d but it failed again. So, I reverted it with https://github.com/up9inc/mizu/commit/20cf3ac471cac931cef60dc739710f7b18afb618

**Comment by gadotroee** [comment]
> > @mertyildiran, can this change help with the acceptance tests or is it not related?
> > I will take a look later today or tomorrow morning (I know it waited for some time)
> > I see that you ran the acceptance tests but it will not run the AMQP acceptance because of this (https://github.com/up9inc/mizu/blob/develop/acceptanceTests/extensions_test.go#L113). We disabled it in the past when they were not stable - so if it can help with the tests we might need to remove this
> 
> I enabled the acceptance tests for AMQP with [c60e9cf](https://github.com/up9inc/mizu/commit/c60e9cf747c0dccc45345790c2c9f49311a0278d) but it failed again. So, I reverted it with [20cf3ac](https://github.com/up9inc/mizu/commit/20cf3ac471cac931cef60dc739710f7b18afb618)

so probably the change here is not the fix for the acceptance test issue, it might be a "new feature" of implementing the requests response matcher for the AMQP, right?  
I knew that we have issue with AMQP and thought it will help with them but maybe it is not the correct assumption

**Comment by mertyildiran** [comment]
> > > @mertyildiran, can this change help with the acceptance tests or is it not related?
> > > I will take a look later today or tomorrow morning (I know it waited for some time)
> > > I see that you ran the acceptance tests but it will not run the AMQP acceptance because of this (https://github.com/up9inc/mizu/blob/develop/acceptanceTests/extensions_test.go#L113). We disabled it in the past when they were not stable - so if it can help with the tests we might need to remove this
> > 
> > 
> > I enabled the acceptance tests for AMQP with [c60e9cf](https://github.com/up9inc/mizu/commit/c60e9cf747c0dccc45345790c2c9f49311a0278d) but it failed again. So, I reverted it with [20cf3ac](https://github.com/up9inc/mizu/commit/20cf3ac471cac931cef60dc739710f7b18afb618)
> 
> so probably the change here is not the fix for the acceptance test issue, it might be a "new feature" of implementing the requests response matcher for the AMQP, right? I knew that we have issue with AMQP and thought it will help with them but maybe it is not the correct assumption

The PR is not aiming to fix the acceptance tests issue. It's adding request-response matching for AMQP such that the requests are matched to their corresponding `Ok`(s) if there are any.

**Comment by leon-up9** [comment]
why the ui-common changes?

**Comment by mertyildiran** [comment]
> why the ui-common changes?

@leon-up9 please see the PR description: https://github.com/up9inc/mizu/pull/1091#issue-1238147684

**Comment by gadotroee** [comment]
I tried to run it in my cluster to get some feelings with this.
I see that all of my requests are response-less, and I can still see the TAB of resopnse (I see only "connection start",  "connection tune", "basic deliver", and "basic publish"

In your screenshot I see the responses on the "basic cancel" "basic consume" and "queue declare" so maybe it is not problem that on all of my requests I don't see response - but it is weird to see tab with no details 

image

image

image


Do you think it is problem in my cluster or maybe those types of requests do not have responses?

File: tap/extensions/amqp/helpers.go (hunk 1 above)
Comment (gadotroee) [reply tap/extensions/amqp/helpers.go]: I see you added logs but they are INFO.
I guess this log should not happen so if in some case we get to this line it is probably error?

**Comment by gadotroee** [comment]
> In your screenshot I see the responses on the "basic cancel" "basic consume" and "queue declare" so maybe it is not problem that on all of my requests I don't see response - but it is weird to see tab with no details

So I think that in this case we need to disable (grey out) or even not show at all this tab if it is possible (feels better than giving the user empty tab)

**Comment by mertyildiran** [comment]
> > In your screenshot I see the responses on the "basic cancel" "basic consume" and "queue declare" so maybe it is not problem that on all of my requests I don't see response - but it is weird to see tab with no details
> 
> So I think that in this case we need to disable (grey out) or even not show at all this tab if it is possible (feels better than giving the user empty tab)

@gadotroee `channel open` method has both an empty request and an empty response. Trying to hide the empty tabs cause problems in the front-end code:

```diff
diff --git a/ui-common/src/components/EntryDetailed/EntryViewer/AutoRepresentation.tsx b/ui-common/src/components/EntryDetailed/EntryViewer/AutoRepresentation.tsx
index 4295b5fa5..7a4a47df6 100644
--- a/ui-common/src/components/EntryDetailed/EntryViewer/AutoRepresentation.tsx
+++ b/ui-common/src/components/EntryDetailed/EntryViewer/AutoRepresentation.tsx
@@ -25,14 +25,20 @@ export const AutoRepresentation: React.FC<any> = ({ representation, color, opene
 
     const { request, response } = JSON.parse(representation);
 
+    console.log(request);
+    console.log(response);
+
     const TABS = useMemo(() => {
-        const arr = [
-            {
+        const arr = []
+
+        if (request.length > 0) {
+            arr.push({
                 tab: 'Request',
                 badge: null
-            }]
+            });
+        }
 
-        if (response) {
+        if (response.length > 0) {
             arr.push({
                 tab: 'Response',
                 badge: null
@@ -40,9 +46,9 @@ export const AutoRepresentation: React.FC<any> = ({ representation, color, opene
         }
 
         return arr
-    }, [response]);
+    }, [request, response]);
 
-    const [currentTab, setCurrentTab] = useState(TABS[0].tab);
+    const [currentTab, setCurrentTab] = useState(TABS[0].tab); // `TABS` expected to have at least one element
 
     const getOpenedTabIndex = useCallback(() => {
         const currentIndex = TABS.findIndex(current => current.tab === currentTab)
@@ -51,7 +57,7 @@ export const AutoRepresentation: React.FC<any> = ({ representation, color, opene
 
     useEffect(() => {
         if (openedTab) {
-            setCurrentTab(TABS[openedTab].tab)
+            setCurrentTab(TABS[openedTab].tab) // `TABS` expected to have at least one element
         }
 
         // eslint-disable-next-line react-hooks/exhaustive-deps
@@ -62,16 +68,17 @@ export const AutoRepresentation: React.FC<any> = ({ representation, color, opene
         return <React.Fragment></React.Fragment>;
     }
 
+    // Can't set `currentTab`
     return <div className={styles.Entry}>
         {<div className={styles.body}>
             <div className={styles.bodyHeader}>
                 <Tabs tabs={TABS} currentTab={currentTab} color={color} onChange={setCurrentTab} leftAligned />
                 {isReplayDisplayed() && <span title="Replay Request"><ReplayIcon fill={color} stroke={color} style={{ marginLeft: "10px", cursor: "pointer", height: "22px" }} onClick={() => setIsOpenRequestModal(true)} /></span>}
             </div>
-            {getOpenedTabIndex() === TabsEnum.Request && <React.Fragment>
+            {request.length > 0 && getOpenedTabIndex() === TabsEnum.Request && <React.Fragment>
                 <SectionsRepresentation data={request} color={color} requestRepresentation={request} />
             </React.Fragment>}
-            {response && getOpenedTabIndex() === TabsEnum.Response && <React.Fragment>
+            {response.length > 0 && getOpenedTabIndex() === TabsEnum.Response && <React.Fragment>
                 <SectionsRepresentation data={response} color={color} />
             </React.Fragment>}
         </div>}
```

**Comment by gadotroee** [comment]
> ```diff
> / `TABS` expected to have at least
> ```

Ok I see, 
it make sense that we need to see the response so I asked @haiut  and we think that in case that we don't have any information to show we can add some text to say "This request has no response data" or something like this so at least this response tab will not be empty.

Can you do it?

**Comment by mertyildiran** [comment]
> > ```diff
> > / `TABS` expected to have at least
> > ```
> 
> Ok I see, it make sense that we need to see the response so I asked @haiut and we think that in case that we don't have any information to show we can add some text to say "This request has no response data" or something like this so at least this response tab will not be empty. (maybe need to be added to request to)
> 
> Can you do it?

Done with https://github.com/up9inc/mizu/commit/112cc65fb62b08c9ab6a9432427af52f545f0fca

**Pull request merged**